### PR TITLE
[NEAT-586] 😧 As parent name

### PR DESCRIPTION
**Pull request by doctrino**
image

### CODE REVIEW

File: cognite/neat/_client/_api/data_modeling_loaders.py [hunk 1]
```diff
@@ -341,19 +341,19 @@ def _retrieve_recursive(
             include_connections: Whether to include all connected views.
             include_ancestors: Whether to include all ancestors.
         """
-        last_batch = list(view_ids)
+        last_batch = set(view_ids)
```
Comment (doctrino) [inline cognite/neat/_client/_api/data_modeling_loaders.py:344]: bug causing duplicates when calling `neat.loaders.views.retrieve(...)`

File: cognite/neat/_session/_mapping.py [hunk 2]
```diff
@@ -37,3 +53,81 @@ def classic_to_core(self, org_name: str) -> None:
         )
 
         self._state.data_model.write(output.rules, change)
+
+        start = datetime.now(timezone.utc)
+
+        source_id, rules = self._state.data_model.last_verified_dms_rules
```
Comment (doctrino) [inline cognite/neat/_session/_mapping.py:59]: Note now that there are three transformations that are done by this method
1. Mapping
2. Including all referenced views and containers.
3. Renaming to parent name.

Will move this in a refactoring.

File: cognite/neat/_rules/transformers/_mapping.py [hunk 3]
```diff
@@ -208,3 +211,128 @@ def _find_overwrites(self, prop: DMSProperty, mapping_prop: DMSProperty) -> tupl
...
+                        continue
+                    container_entity = ContainerEntity.from_id(property_.container)
+                    view_by_container_properties[(container_entity, property_.container_property_identifier)].append(
+                        (view_entity, property_id)
+                    )
+
+        return view_by_container_properties
+
+    @staticmethod
+    def _get_parent_name_by_container_property(
```
Comment (nikokaoja) [inline cognite/neat/_rules/transformers/_mapping.py:328]: I guess you mean `_get_parent_view_property....`?

File: cognite/neat/_rules/transformers/_mapping.py [hunk 4]
```diff
@@ -208,3 +211,128 @@ def _find_overwrites(self, prop: DMSProperty, mapping_prop: DMSProperty) -> tupl
...
+
+    def transform(self, rules: DMSRules | OutRules[DMSRules]) -> JustRules[DMSRules]:
+        input_rules = self._to_rules(rules)
+        new_rules = input_rules.model_copy(deep=True)
+        new_rules.metadata.version += "_as_parent_name"
+
+        path_by_view = self._path_by_view(new_rules)
+        view_by_container_property = self._view_by_container_properties(new_rules)
+
+        parent_name_by_container_property = self._get_parent_name_by_container_property(
```
Comment (nikokaoja) [inline cognite/neat/_rules/transformers/_mapping.py:232]: see my comment on how to call this method as this is not very descriptive name

File: cognite/neat/_rules/transformers/_mapping.py [hunk 5]
```diff
@@ -208,3 +211,128 @@ def _find_overwrites(self, prop: DMSProperty, mapping_prop: DMSProperty) -> tupl
                     # These are used for warnings so we use the alias to make it more readable for the user
                     conflicts.append(mapping_prop.model_fields[field_name].alias or field_name)
         return to_overwrite, conflicts
+
+
+class AsParentName(RulesTransformer[DMSRules, DMSRules]):
+    """Looks up all properties that maps to the same container property,
```
Comment (nikokaoja) [inline cognite/neat/_rules/transformers/_mapping.py:217]:
```suggestion
    """Looks up all view properties that map to the same container property,
```

File: cognite/neat/_rules/transformers/_mapping.py [hunk 6]
```diff
@@ -208,3 +211,128 @@ def _find_overwrites(self, prop: DMSProperty, mapping_prop: DMSProperty) -> tupl
...
+    and changes the child property name to match the parent property name.
```
Comment (nikokaoja) [inline cognite/neat/_rules/transformers/_mapping.py:218]:
```suggestion
    and changes the child view property name to match the parent property name.
```

File: cognite/neat/_rules/transformers/_mapping.py [hunk 7]
```diff
@@ -208,3 +211,128 @@ def _find_overwrites(self, prop: DMSProperty, mapping_prop: DMSProperty) -> tupl
                     # These are used for warnings so we use the alias to make it more readable for the user
                     conflicts.append(mapping_prop.model_fields[field_name].alias or field_name)
         return to_overwrite, conflicts
+
+
+class AsParentName(RulesTransformer[DMSRules, DMSRules]):
```
Comment (nikokaoja) [inline cognite/neat/_rules/transformers/_mapping.py:216]: We are not changing property name but property id, so suggest calling this 
`AsParentPropertyId` as that is what we are doing here

File: cognite/neat/_session/_mapping.py [hunk 8]
```diff
@@ -1,28 +1,44 @@
...
+from cognite.neat._store._provenance import Agent as ProvenanceAgent
 from cognite.neat._store._provenance import Change
 
 from ._state import SessionState
-from .exceptions import session_class_wrapper
+from .exceptions import NeatSessionError, session_class_wrapper
 
 
 @session_class_wrapper
 class MappingAPI:
```
Comment (nikokaoja) [inline cognite/neat/_session/_mapping.py:17]: to keep our interfaces/interaction with NeatSession uniform, MappingAPI should have split among mappings done on instances and mappings done on data model:

File: cognite/neat/_session/_mapping.py [hunk 9]
```diff
@@ -1,28 +1,44 @@
...
-    def __init__(self, state: SessionState):
+    def __init__(self, state: SessionState, client: NeatClient | None = None):
         self._state = state
+        self._client = client
```
Comment (nikokaoja) [inline cognite/neat/_session/_mapping.py:20]: I strongly suggest here
```suggestion
        self._client = client
        self.data_model = DataModelMappingAPI(self._state, self._client)
```

File: cognite/neat/_session/_mapping.py [hunk 10]
```diff
@@ -1,28 +1,44 @@
...
 
-    def classic_to_core(self, org_name: str) -> None:
+    def classic_to_core(self, company_prefix: str, use_parent_property_name: bool = True) -> None:
```
Comment (nikokaoja) [inline cognite/neat/_session/_mapping.py:22]: move all this under `DataModelMappingAPI`

**Review by nikokaoja — APPROVED** [review]
Approving as you will prob move most of my feedbacks to dedicated Jira issues

File: cognite/neat/_rules/transformers/_mapping.py [hunk 11]
```diff
@@ -208,3 +211,128 @@ def _find_overwrites(self, prop: DMSProperty, mapping_prop: DMSProperty) -> tupl
...
+
+        # We need to look up all parent properties.
+        to_lookup = {view.view.as_id() for view in rules.views if view.view not in view_with_properties}
+        if to_lookup and self._client is None:
+            raise CDFMissingClientError(
+                f"Views {to_lookup} are not in the data model. Please provide a client to lookup the views."
+            )
+        elif to_lookup and self._client:
+            read_views = self._client.loaders.views.retrieve(list(to_lookup), include_ancestor=True)
+            write_views = [self._client.loaders.views.as_write(read_view) for read_view in read_views]
```
[the remark this answers is not in the document]
Comment (doctrino) [reply cognite/neat/_rules/transformers/_mapping.py:309]: Marking this as a task for later.

**Pull request merged**